### ISSUE TITLE
MessageVerifier's expires_in and expires_at options can yield different expiries

### ISSUE THREAD

**Issue by shouichi** [rails/rails#44808]
### Steps to reproduce

`ActiveSupport::MessageVerifier#generate` has two similar options: `expires_in` and `expires_at`. But they can yield different expiries when passing `expires_in: 1.month` or `expires_at: 1.month.from_now`.

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "activesupport"
end

require "active_support"
require "active_support/core_ext"
require "active_support/testing/time_helpers"
require "minitest/autorun"

class BugTest < Minitest::Test
  include ActiveSupport::Testing::TimeHelpers

  def setup
    @verifier = ActiveSupport::MessageVerifier.new("secret")
    @edge_case_time = Time.parse("2022-03-30T23:59:59Z")
  end

  def test_expires_in
    message = travel_to(@edge_case_time) do
      # Internally, it creates ActiveSupport::Messages::Metadata with
      # @expire_at 2022-04-30T23:59:59.000Z
      @verifier.generate("whatever", expires_in: 1.month)
    end

    travel_to(@edge_case_time + 1.month - 1.second) do
      assert_raises ActiveSupport::MessageVerifier::InvalidSignature do
        @verifier.verify(message)
      end
    end
  end

  def test_expires_at
    message = travel_to(@edge_case_time) do
      # Internally, it creates ActiveSupport::Messages::Metadata with
      # @expire_at 2022-04-29T23:59:59.000Z
      @verifier.generate("whatever", expires_at: 1.month.from_now)
    end

    travel_to(@edge_case_time + 1.month - 1.second) do
      assert_raises ActiveSupport::MessageVerifier::InvalidSignature do
        @verifier.verify(message)
      end
    end
  end
end

```

### Expected behavior

Both `expires_in: 1.month` and `expires_at: 1.month.from_now` yield the same expiry.

### Actual behavior

When the time is 2022-03-30T23:59:59Z, expiries are 

- `expires_in`: 2022-04-30T23:59:59.000Z
- `expires_at`: 2022-04-29T23:59:59.000Z

I believe this commit https://github.com/rails/rails/commit/9fbfb4bffe72320cc895e2f4eb4e30e108d58bb2 introduced this behavior.

### System configuration
**Rails version**: 7.0.2.3

**Ruby version**: 3.1.1

**Comment by SkipKayhil** [comment]
Hey @shouichi, I tried running the script on my machine and haven't been able to reproduce the test failures. Could it be an issue with the device or ruby installation?

My ruby is `ruby 3.1.1p18 (2022-02-18 revision 53f5fc4236) [x86_64-linux]` running inside WSL2.

**Comment by shouichi** [comment]
Mmm, you're right. The tests pass now. But I'm pretty sure it happened. I'll try to reproduce it. Thank you.

**Comment by shouichi** [comment]
@SkipKayhil Changed the reproduction script and I believe it now fails.

I believe this behavior was introduced in this commit https://github.com/rails/rails/commit/9fbfb4bffe72320cc895e2f4eb4e30e108d58bb2.

Thank you.

**Comment by jonathanhefner** [comment]
Hi, @shouichi! :wave:  I believe your reproduction script is not correct.  The message should not be expired at time `@edge_case_time + 1.month - 1.second`, so `@verifier.verify` should not raise.  I believe your original script (which traveled to `1.month.from_now + 1.second`) was correct, but I could not reproduce the bug with it.

Here is another version of the script that I believe is correct, but also does not reproduce the bug:

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "activesupport", "~> 7.0.0"
end

require "active_support"
require "active_support/core_ext"
require "active_support/testing/time_helpers"
require "minitest/autorun"

class BugTest < Minitest::Test
  include ActiveSupport::Testing::TimeHelpers

  def setup
    @verifier = ActiveSupport::MessageVerifier.new("secret")
    @edge_case_time = Time.parse("2022-03-30T23:59:59Z")
  end

  def test_expires_in
    data = "whatever"
    message = travel_to(@edge_case_time) do
      @verifier.generate(data, expires_in: 1.month)
    end

    travel_to(@edge_case_time + 1.month - 1.second) do
      assert_equal data, @verifier.verified(message)
    end

    travel_to(@edge_case_time + 1.month + 1.second) do
      assert_nil @verifier.verified(message)
    end
  end

  def test_expires_at
    data = "whatever"
    message = travel_to(@edge_case_time) do
      @verifier.generate(data, expires_at: 1.month.from_now)
    end

    travel_to(@edge_case_time + 1.month - 1.second) do
      assert_equal data, @verifier.verified(message)
    end

    travel_to(@edge_case_time + 1.month + 1.second) do
      assert_nil @verifier.verified(message)
    end
  end
end
```

If you can reproduce the bug with that script (change `@edge_case_time`, if necessary), please let me know.

**Comment by shouichi** [comment]
@jonathanhefner Thank you for the response! I fixed the script. The changes I made were

- Make sure to use a certain timezone (the bug is timezone independent).
- Fix the time to call `verity` (the problem is that the expiry is set earlier than expected).

Hope this reproduces the bug 🙏

```
require "bundler/inline"

gemfile(false) do
  source "https://rubygems.org"

  gem "activesupport", path: "./activesupport"
end

require "active_support"
require "active_support/core_ext"
require "active_support/testing/assertions"
require "active_support/testing/time_helpers"
require "minitest/autorun"

class BugTest < Minitest::Test
  include ActiveSupport::Testing::Assertions
  include ActiveSupport::Testing::TimeHelpers

  def setup
    Time.zone = "Asia/Tokyo"
    @verifier = ActiveSupport::MessageVerifier.new("secret")
    @edge_case_time = Time.parse("2022-03-30T23:59:59Z")
  end

  def test_expires_in
    message = travel_to(@edge_case_time) do
      # Internally, it creates ActiveSupport::Messages::Metadata with
      # @expire_at 2022-04-30T23:59:59.000Z
      @verifier.generate("whatever", expires_in: 1.month)
    end

    travel_to(@edge_case_time + 1.month - 1.day) do
      assert_nothing_raised { @verifier.verify(message) }
    end
  end

  def test_expires_at
    message = travel_to(@edge_case_time) do
      # Internally, it creates ActiveSupport::Messages::Metadata with
      # @expire_at 2022-04-29T23:59:59.000Z
      @verifier.generate("whatever", expires_at: 1.month.from_now)
    end

    travel_to(@edge_case_time + 1.month - 1.day) do
      assert_nothing_raised { @verifier.verify(message) }
    end
  end
end
```

**Comment by jonathanhefner** [comment]
Your reproduction script indicates that the problem is in the `test_expires_at` test.  Therefore, #45472 will not fix it, because #45472 does not affect `expires_at`.

Digging deeper, this actually appears to be an issue with `ActiveSupport::TimeWithZone#advance`:

* `from_now` is an [alias of `since`](https://github.com/rails/rails/blob/5cf939ff18aeab77295a7c86027f68d0bb5d578e/activesupport/lib/active_support/duration.rb#L433), so `1.month.from_now` == `1.month.since(Time.current)`.
* `since` [calls `Duration#sum`](https://github.com/rails/rails/blob/5cf939ff18aeab77295a7c86027f68d0bb5d578e/activesupport/lib/active_support/duration.rb#L431).
* `Duration#sum` [calls `ActiveSupport::TimeWithZone#advance`](https://github.com/rails/rails/blob/5cf939ff18aeab77295a7c86027f68d0bb5d578e/activesupport/lib/active_support/duration.rb#L496).
* `ActiveSupport::TimeWithZone#advance` [eventually](https://github.com/rails/rails/blob/5cf939ff18aeab77295a7c86027f68d0bb5d578e/activesupport/lib/active_support/time_with_zone.rb#L455) makes a [call to `Time#advance`](https://github.com/rails/rails/blob/5cf939ff18aeab77295a7c86027f68d0bb5d578e/activesupport/lib/active_support/time_with_zone.rb#L570).
* Because months can have a different number of days (28, 29, 30, or 31), `Time#advance` [changes](https://github.com/rails/rails/blob/5cf939ff18aeab77295a7c86027f68d0bb5d578e/activesupport/lib/active_support/core_ext/time/calculations.rb#L193-L194) `2022-03` to `2022-04`.  `2022-04-31` does not exist, so it becomes `2022-04-30`.

Reduced reproduction:

```ruby
utc_time = Time.find_zone!("UTC").parse("2022-03-30T23:59:59Z")
# => Wed, 30 Mar 2022 23:59:59.000000000 UTC +00:00

utc_time.advance(months: 1)
# => Sat, 30 Apr 2022 23:59:59.000000000 UTC +00:00

jst_time = Time.find_zone!("Asia/Tokyo").parse("2022-03-30T23:59:59Z")
# => Thu, 31 Mar 2022 08:59:59.000000000 JST +09:00

jst_time.advance(months: 1)
# => Sat, 30 Apr 2022 08:59:59.000000000 JST +09:00

jst_time.time
# => 2022-03-31 08:59:59 UTC

jst_time.time.advance(months: 1)
# => 2022-04-30 08:59:59 UTC
```

I am not sure of the best solution.

**Comment by shouichi** [comment]
@jonathanhefner Sorry for replying late. You are right. The PR did not fix
the problem and I closed the one.

I think I found the root problem. The problem is that `TimeWithZone` handles
timezone incorrectly when adding duration. Here's an example.

```
> jst = Time.find_zone!("Asia/Tokyo").parse("2022-03-30T23:59:59Z")
> jst.utc + 1.month
=> 2022-04-30 23:59:59 UTC # correct
> (jst + 1.month).utc
=> 2022-04-29 23:59:59 UTC # wrong
```

It seems like `TimeWithZone` converts itself to UTC, advances by duration then
converts back to the original timezone.

https://github.com/rails/rails/blob/e240daa6cc013b5a24fb692a8b82ce18dc44ad11/activesupport/lib/active_support/time_with_zone.rb#L570

However, this `time` variable seems wrong. It is UTC but timezone offset is not
added/subtracted. Here's an example.

```
> jst = Time.find_zone!("Asia/Tokyo").parse("2022-03-30T23:59:59Z")
> jst.utc
=> 2022-03-30 23:59:59 UTC # correct
> jst.time
=> 2022-03-31 08:59:59 UTC # wrong
```
